Library/OcSmbiosLib/SmbiosPatch.c

```diff
@@ -1017,26 +1033,45 @@ PatchBootInformation (
 
   @param[in] Table                  Pointer to location containing the current address within the buffer.
   @param[in] Data                   Pointer to tocation containing SMBIOS data.
+  @param[in] AppleSMBIOS            TRUE if Apple SMBIOS present, FALSE otherwise.
 **/
 STATIC
 VOID
-CreateAppleFirmwareVolume (
+CreateOrPatchAppleFirmwareVolume (
   IN OUT OC_SMBIOS_TABLE  *Table,
-  IN     OC_SMBIOS_DATA   *Data
+  IN     OC_SMBIOS_DATA   *Data,
+  IN     BOOLEAN          AppleSMBIOS
   )
 {
+  APPLE_SMBIOS_STRUCTURE_POINTER  Original;
   UINT8                           MinLength;
 
+  UINT32 FirmwareFeatures;
+  UINT32 FirmwareFeaturesMask;
+  UINT32 ExtendedFirmwareFeatures;
+  UINT32 ExtendedFirmwareFeaturesMask;
+
+  if (AppleSMBIOS) {
+    Original      = SmbiosGetOriginalStructure (APPLE_SMBIOS_TYPE_FIRMWARE_INFORMATION, 1);
+  } else {
+    Original.Raw  = NULL;
+  }
+
   MinLength   = sizeof (*Table->CurrentPtr.Type128);
 
   if (EFI_ERROR (SmbiosInitialiseStruct (Table, APPLE_SMBIOS_TYPE_FIRMWARE_INFORMATION, MinLength, 1))) {
     return;
   }
 
-  Table->CurrentPtr.Type128->FirmwareFeatures             = (UINT32)BitFieldRead64 (Data->FirmwareFeatures,     0,  31);
-  Table->CurrentPtr.Type128->FirmwareFeaturesMask         = (UINT32)BitFieldRead64 (Data->FirmwareFeaturesMask, 0,  31);
-  Table->CurrentPtr.Type128->ExtendedFirmwareFeatures     = (UINT32)BitFieldRead64 (Data->FirmwareFeatures,     32, 63);
-  Table->CurrentPtr.Type128->ExtendedFirmwareFeaturesMask = (UINT32)BitFieldRead64 (Data->FirmwareFeaturesMask, 32, 63);
+  FirmwareFeatures             = (UINT32)BitFieldRead64 (Data->FirmwareFeatures    ,  0, 31);
```

Comment:
A space should also be followed after casting just like calling functions. Please fix these lines.